### ISSUE TITLE
Customer Accounts

### ISSUE THREAD

**Issue by midsoft2010** [opensourcepos/opensourcepos#3072]
### Background information

__IMPORTANT: If you choose to ignore this issue report template, your issue will be closed as we cannot help without the requested information.__

Please make sure you tick (add an x between the square brackets with no spaces) the following check boxes:

- [] Reporting an issue of an unmodified OSPOS installation
- [] Checked [open and closed issues](https://github.com/opensourcepos/opensourcepos/issues?utf8=%E2%9C%93&q=is%3Aissue) and no similar issue was already reported (please make sure you searched!)
- [] Read [README](https://github.com/opensourcepos/opensourcepos/blob/master/README.md), [WHATS_NEW](https://github.com/opensourcepos/opensourcepos/blob/master/WHATS_NEW.txt), [INSTALL.md](https://github.com/opensourcepos/opensourcepos/blob/master/INSTALL.md) and [UPGRADE](https://github.com/opensourcepos/opensourcepos/blob/master/UPGRADE.txt)
- [] Read the [FAQ](https://github.com/opensourcepos/opensourcepos#faq) for any known install and/or upgrade gotchas (in specific PHP extensions installed)
- [] Read the [wiki](https://github.com/opensourcepos/opensourcepos/wiki)
- [] Executed any database upgrade scripts if an upgrade pre 3.0.0 (e.g. database/2.4_to_3.0.sql)
- [] Aware the installation code is in [bintray](https://bintray.com/jekkos/opensourcepos/opensourcepos/view/files?sort=updated&order=asc#files) (see README), and [GitHub master](https://github.com/opensourcepos/opensourcepos/tree/master) is for [developers only](https://github.com/opensourcepos/opensourcepos/wiki/Development-setup) and therefore not complete nor stable

### Installation information

- OSPOS version is: 
- OSPOS git commit hash is: 
- PHP version is: 
- MySQL or MariaDB version is: 
- OS and version is: 
- WebServer is: 
- Selected language is: 
- (If applicable) Docker installation: 
- (If applicable) Installation package for the LAMP/LEMP stack is: 

### New Feature
Somebody to improve this owesome pos system by adding customer Accounts where due sales is counted separately and can be cleared later after payments. Don't forget to add credit limit(the amount the customer cant cannot due beyond)

**Comment by SteveIreland** [comment]
The following tickets appear related.  I'm willing to spend some time on these issues so  I'm assigning these to myself.  If someone else is planning on contributing some work they've already done on it please let me know so I don't waste my time.
#2275, #2679, #2045, #2285, #3072

**Comment by SteveIreland** [comment]
Since this is the cleanest of the set of related tickets I will be using this one for updates/questions/etc.

The first change I'll make is to add a new filter to the detail transaction report that will only list sales within the select date range that have a balance due amount.  I'll also add the balance due amount to the list of payments.  The report itself probably has as many columns of information as it can properly handle which is why I'm proposing just adding it as another payment type labeled "Balance" and showing the amount of the balance due.

The second change will be to add a new detail report named Receivables.  This will drop one or two columns and add a column for Balance Due.  Of course the link to payment maintenance for the sale will still be there so that payments can be made after selecting the sale to apply the payment.

The final change will be to rework the payment maintenance screen a bit.   It's already been changed to allow post sale payments to be added, but it's getting a bit cumbersome to use with the recent addition of cash adjustments and post sale refunds.   I'm not sure what the exact changes need to be made, but I'm definitely going to add a column for the date/time when the payments were applied and I'd like to make it a bit more intelligent when handling the transition from a cash only sale to a cash sale that now has a non-cash payment type (where cash rounding is involved).

Please suggest any corrections or additional changes that need to be made.  (Other than adding on a stand-alone accounts receivable and statement generation system ... that's not going to happen.)

**Comment by SteveIreland** [comment]
The following is an example of the first change which I have working at the moment.  This is the Detailed Sales Transaction report.  There IS a balance due filter so that only those sales with a balance due are listed, but it will include the balance due of an individual sale on all flavors of the report.

![image](https://user-images.githubusercontent.com/23066623/108649539-8bd1f100-748b-11eb-8844-37c7a1058b67.png)

**Comment by ozzylamberth** [comment]
I DO NOT believe that at this time the OSPOS collaborators are preparing to carry out the account modules as such. But keeping the accounts payable of the clients is important to have a readable control of the accounts payable in my case it would be assigned in the Client view module as indicated in issue 2679

![capture-20210222-015834](https://user-images.githubusercontent.com/7409604/108668886-5f7d9b00-74b2-11eb-86cc-c773f4dacd68.png)

But without eliminating the total expense column. That is, keep the columns and only assign the column of Amount pending cancellation of a credit, bond or any other type of debt.
and allow the line to be broken down by pending invoices to be canceled. It is serious my idea in the same way it would give rise to the Accounts Receivable modules indirectly. For the credit limits I would not worry about it but rather I would worry about the time that a client must have to cancel. A date limit would be ideal.
Suppose that I am going to make a sale to a frequent client which will be a credit sale with cancellation 15 days after the purchase that would allow me to charge interest for late payment and also limit the sale to the client until the cancellation is made. of bills pending payment.

Of course, in my case I am a Wholesale and Retail Seller

**Comment by SteveIreland** [comment]
Fortunately for me I still have a job and hope to keep it so writing an account management module is out of the question.  I started one but reality slapped me down.

However, I believe I should be able to add the balance due on the register and on the customer list.

@jekkos should this be a configurable feature or should it be part of OSPOS base package?

**Comment by midsoft2010** [comment]
> Since this is the cleanest of the set of related tickets I will be using this one for updates/questions/etc.
> 
> The first change I'll make is to add a new filter to the detail transaction report that will only list sales within the select date range that have a balance due amount. I'll also add the balance due amount to the list of payments. The report itself probably has as many columns of information as it can properly handle which is why I'm proposing just adding it as another payment type labeled "Balance" and showing the amount of the balance due.
> 
> The second change will be to add a new detail report named Receivables. This will drop one or two columns and add a column for Balance Due. Of course the link to payment maintenance for the sale will still be there so that payments can be made after selecting the sale to apply the payment.
> 
> The final change will be to rework the payment maintenance screen a bit. It's already been changed to allow post sale payments to be added, but it's getting a bit cumbersome to use with the recent addition of cash adjustments and post sale refunds. I'm not sure what the exact changes need to be made, but I'm definitely going to add a column for the date/time when the payments were applied and I'd like to make it a bit more intelligent when handling the transition from a cash only sale to a cash sale that now has a non-cash payment type (where cash rounding is involved).
> 
> Please suggest any corrections or additional changes that need to be made. (Other than adding on a stand-alone accounts receivable and statement generation system ... that's not going to happen.)

I think is add something like credit limit is very Important:- assuming that amount due is the balance the customer owes there should be a column where the admin can add limit( due limit). This means if customer "due limit" or  "credit limit" is exhausted user cannot sell using due or the one you are calling balance to that specific customer who limit has exhausted.

**Comment by midsoft2010** [comment]
![Screenshot (29)](https://user-images.githubusercontent.com/29373167/108834182-e5970000-75de-11eb-8c8e-1f83f0a1279e.png)
something similar to this

**Comment by SteveIreland** [comment]
Okay, if no one vetoes this change, then I will add a credit limit amount to the customer table.  If zero then it will be ignored.  If greater than zero then the order entry person will not be able to complete the order if the balance due (unpaid balance) of prior sales for the selected customer PLUS the current order amount is greater than credit credit limit amount.  

If a balance due is greater than zero it will be displayed on the register customer information.

If the customer has a credit limit and the balance due is greater than zero then the credit limit will also be displayed.

**Comment by jekkos** [comment]
@SteveIreland if a customer has a balance due, how will we process this in the receipt? Will we add an additional payment to make up for this balance? Not sure how this is currently handled

**Comment by SteveIreland** [comment]
@jekkos The current payment update panel provides the ability to add a payment against an invoice.  I believe what is currently happening is that the balance due for the sale is computed and the user simply needs to select the payment type.  The user can change the balance due to the amount actually paid (for example if the user was only able to make a partial payment) which will leave a balance due for the sale.   After payment, the sales document can be reprinted and handed or mailed to the customer so that they have a record of the remaining balance on the sale.

The complication that I'll be trying to sort through (and might need some advice from those who use the cash rounding feature) is that if the original sale was partially paid as a cash sale then the amount due is based on cash rounding kicking in.  If they then follow up with a non-cash payment then that will change the original sale amount (at least I think it should).  But since I don't know what actually happens in the real world of cash rounding I'll be guessing at what should happen.  My current guess is that since it is no longer cash only then the government will not allow cash rounding and therefore the balance due should be changed to reflect the non-rounded balance.

**Comment by jekkos** [comment]
> Okay, if no one vetoes this change, then I will add a credit limit amount to the customer table. If zero then it will be ignored. If greater than zero then the order entry person will not be able to complete the order if the balance due (unpaid balance) of prior sales for the selected customer PLUS the current order amount is greater than credit credit limit amount.
> 
> If a balance due is greater than zero it will be displayed on the register customer information.
> 
> If the customer has a credit limit and the balance due is greater than zero then the credit limit will also be displayed.

Sounds good.. but will there be a way to track this outstanding debt? I guess it will be visible in the customers section directly. I'm not sure how this is taken into account in for example the takings screen. Do we add up this amount on credit for the daily statements eg?

**Comment by SteveIreland** [comment]
My though is to include the current outstanding balance on the customer summary and customer detail reports.  I'll also include it on the register and the daily sales.

When you refer to the Takings Screen are you referring to the invoice update where payment types are adjusted and additional payments are made?  On that panel the outstanding balance is already automatically calculated and the user is prompted to enter that payment amount.

What are the daily statements are you referring to?  I didn't know we had daily statements (or at least my understanding of what a statement is).

**Comment by SteveIreland** [comment]
Looked at the customer stats panel and noticed that it only looks at payments to report sales.  So it assumes all sales are paid in full.  In order to repair this situation I will need to replace the query with the more complete version that involves more overhead.

**Comment by SteveIreland** [comment]
I'm making progress on this but I seem to have opened a can of worms by trying to support conventional balance due while also supporting the "Due" payment type.  I can support it but it's going to make things a bit complex ... probably a little more complex than what we want.

I'd like to propose that we change the sales register to allow completing a POS transaction simply by selecting a checkbox labeled "Close with Due" instead of manually entering a Due payment transaction.

I would add a migration step to delete all "Due" payment transactions.  The system will track payment due amounts by simply computing the sale amount minus the net payment (tendered minus refund).

Any objections?

**Comment by SteveIreland** [comment]
Question.  Rewards are currently being updated based on payment transactions.  For POS sales this included Due transactions.  For Invoices, since there isn't a Due payment transaction, rewards are only updated based on payments received.

Should rewards be updated with the total sales amount or just payment received at the point of sale?

**Comment by truchosky** [comment]
my vote is just payment received

**Comment by truchosky** [comment]
hello, any news about this?, I was specting this long time ago

**Comment by midsoft2010** [comment]
Also waiting for any good news about this

**Comment by hutgold** [comment]
I'd like to bump this, it seems like the devs had a working version of this and I would love to see it in action, 
Thank you!

**Comment by Monstro30** [comment]
Hi everyone, is there any update to the debt tracking (DUE) function ?

**Comment by SteveIreland** [comment]
All I can say about this, is that I thought I was off permanently off the project.  But life is unpredictable and I'm back.

Thank you for reminding me that I had committed to move this forward.

This is something my client needs (although they haven't yet requested it).  Fortunately for me, their radar is underpowered.

Their last request was to add a module for appointment and work order scheduling - so that will be first (and will only be added to my fork).  Then I'll start working on the Customer Accounts feature.  That  needs to happen within the next few months.   I'll probably add that to my fork first.  As soon as the CI4 branch gets merged to master, the first thing I'll do is add the Customer Accounts feature to base OSPOS, which should only take a weekend or two.

**Comment by midsoft2010** [comment]
Still waiting for this!

**Comment by SteveIreland** [comment]
And I believe I'm still waiting on the CI4 branch to master.

**Comment by hoststatic** [comment]
Any new update about this feature @SteveIreland

**Comment by SteveIreland** [comment]
I've been overly busy with paid work for awhile, but I believe for this effort I'm still waiting on the CI4 branch to be pushed to master.   Hopefully the technology experts are getting close to making it happen.  As soon as it happens then I'll start work on this, as well as migrating my only OSPOS customer (with their customizations) to CI4.
